USM allocator won't compile on macOS with std::allocate_shared

**Bug summary**
The following code fails to compile (on macOS at least) with the current branch of hipSYCL.
It seems to come from this copy constructor:
```{C++}
 template <class U>
  usm_allocator(const usm_allocator<U, AllocKind, Alignment> &other) noexcept
      : _ctx{other._ctx}, _dev{other._dev} {}
```

In usm.hpp, line 333, declaring `usm_allocator` as a friend with 
```
private:
  template <class U, usm::alloc AllocKindU, size_t AlignmentU>
  friend class usm_allocator;
  context _ctx;
  device _dev;
```
solves the issue (see also Intel/llvm implementation of `usm_allocator`).

**To Reproduce**
```
#include <sycl/sycl.hpp>

template<typename T>
std::shared_ptr<T> get_shared(T &&val, sycl::queue q) {
    typedef sycl::usm_allocator<T, sycl::usm::alloc::shared> shared_allocator;
    return std::allocate_shared<T, shared_allocator>(shared_allocator(q), val);
}

int main() {
    sycl::queue q;
    auto ptr = get_shared(100, q);
    return 0;
}
```
fails with the following: 


```
In file included from usm_allocator_test.cpp:1:
In file included from /usr/local/bin/../include/sycl/sycl.hpp:31:
In file included from /usr/local/bin/../include/sycl/../hipSYCL/sycl/sycl.hpp:82:
/usr/local/bin/../include/sycl/../hipSYCL/sycl/usm.hpp:303:20: error: '_ctx' is a private member of 'hipsycl::sycl::usm_allocator<int, hipsycl::sycl::usm::alloc::shared, 0>'
      : _ctx{other._ctx}, _dev{other._dev} {}
                   ^
/Library/Developer/CommandLineTools/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/memory:4059:9: note: in instantiation of function template specialization 'hipsycl::sycl::usm_allocator<std::__1::__shared_ptr_emplace<int, hipsycl::sycl::usm_allocator<int, hipsycl::sycl::usm::alloc::shared, 0>>, hipsycl::sycl::usm::alloc::shared, 0>::usm_allocator<int>' requested here
    _A2 __a2(__a);
        ^
usm_allocator_test.cpp:7:17: note: in instantiation of function template specialization 'std::__1::allocate_shared<int, hipsycl::sycl::usm_allocator<int, hipsycl::sycl::usm::alloc::shared, 0>, int &>' requested here
    return std::allocate_shared<T, shared_allocator>(shared_allocator(q), val);
                ^
usm_allocator_test.cpp:11:16: note: in instantiation of function template specialization 'get_shared<int>' requested here
    auto ptr = get_shared(100);
               ^
/usr/local/bin/../include/sycl/../hipSYCL/sycl/usm.hpp:334:11: note: declared private here
  context _ctx;
          ^
/usr/local/bin/../include/sycl/../hipSYCL/sycl/usm.hpp:303:38: error: '_dev' is a private member of 'hipsycl::sycl::usm_allocator<int, hipsycl::sycl::usm::alloc::shared, 0>'
      : _ctx{other._ctx}, _dev{other._dev} {}
                                     ^
/usr/local/bin/../include/sycl/../hipSYCL/sycl/usm.hpp:335:10: note: declared private here
  device _dev;
         ^
2 errors generated.
gmake[3]: *** [CMakeFiles/usm_allocator_test.dir/build.make:84: CMakeFiles/usm_allocator_test.dir/usm_allocator_test.cpp.o] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:101: CMakeFiles/usm_allocator_test.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:108: CMakeFiles/usm_allocator_test.dir/rule] Error 2
gmake: *** [Makefile:137: usm_allocator_test] Error 2
```

--- a/include/hipSYCL/sycl/usm.hpp
+++ b/include/hipSYCL/sycl/usm.hpp
@@ -331,6 +331,8 @@ public:
   }
 
 private:
+  template <class U, usm::alloc AllocKindU, size_t AlignmentU>
+  friend class usm_allocator;
   context _ctx;
   device _dev;
 };

